### PR TITLE
Initial implementation of DNASNode

### PR DESCRIPTION
**Pull request by jkarns275**
Creates a node that can use differentiable neural architecture techniques to learn the best node type at a given location. In particular, the node will learn a [Gumbel-Softmax distribution](https://arxiv.org/pdf/1611.01144.pdf), as is done in [this paper](https://arxiv.org/pdf/1812.09926.pdf).

Basically, a DNASNode is supplied a set of nodes and will learn a set of weights (pi) to apply to the node outputs during backpropagation. These weights can be used as a proxy to node importance.

I still need to add a gradient test before I create a mutation that uses this operation, but I have compared the formulas I used to the gradients pytorch generates so they should be correct.

Once I have a mutation for this, I can start to play with the hyperparameters and policies. Some of these include:
- Temperature schedule
- Policy to determine when we can make a conclusion about which node is best (e.g. after X iterations training iterations)
- What to do after a conclusion is made, some sort of crystallization step

### CODE REVIEW

**Comment by jkarns275** [comment]
rfc @travisdesell @zimenglyu

**Review by travisdesell — APPROVED** [review]
